File: java/ems_negocio/ems_negocio-ejb/ejbModule/br/erlangms/service/EmsServiceFacade.java
/*********************************************************************
 * @title Módulo EmsServiceFacade
 * @version 1.0.0
 * @doc Session Bean implementation class EmsServiceFacade
 * @author Everton de Vargas Agilar <evertonagilar@gmail.com>
 * @copyright ErlangMS Team
 *********************************************************************/
 
package br.erlangms.service;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;

import br.erlangms.EmsAgent;


/**
 * Classe base para serviços ErlangMS
 */
public abstract class EmsServiceFacade extends EmsAgent {
	public enum States {BEFORESTARTED, STARTED, PAUSED, SHUTTINGDOWN};
    private States state;
       	
    @PostConstruct
    public void initialize() throws Exception {
        state = States.BEFORESTARTED;
        print_log("Carregando...");
        start();
        state = States.STARTED;
        print_log("Carregado!");
    }
    
    @PreDestroy
    public void terminate() {
        state = States.SHUTTINGDOWN;
        print_log("Finalizando...");
        close();
        print_log("Finalizado.");
    }

    public States getState() {
        return state;
    }
    
    public void setState(States state) {
        this.state = state;
    }    
}

File: berco/Cliente/Cliente/EmsServiceFacade.cs
using System;
namespace br.erlangms
{
	public abstract class EmsServiceFacade : IEmsServiceFacade
	{
		private EmsConnection connection = null;
		private DaemonThread daemon = null;
		public enum States { BEFORESTARTED, STARTED, PAUSED, SHUTTINGDOWN };
		private States state;

		public void initialize()
		{
			state = States.BEFORESTARTED;
			Class <? extends EmsServiceFacade > cls = getClass();
			connection = new EmsConnection(cls.getSimpleName(), cls.getName(), this);
			daemon = new DaemonThread(connection);
			daemon.start();
			state = States.STARTED;
		}

		public void terminate()
		{
			state = States.SHUTTINGDOWN;
			daemon.stop();
			connection.close();
			connection = null;
		}

		protected States getState()
		{
			return state;
		}

		protected EmsConnection getConnection()
		{
			return connection;
		}

		private class DaemonThread : Thread
		{

			private EmsConnection agent;

			public DaemonThread(EmsConnection agent) : base(agent)
			{
				this.agent = agent;
			}

			public void run()

			{
				try
				{
					agent.start();
				}
				catch (Exception e)
				{
					Console.WriteLine(e.ToString());
				}
			}
		}
	}
}